### PR TITLE
Fix invalid session secure cookie setting

### PR DESCRIPTION
**Pull request by wsargent**
The configuration value should be "play.http.session.secure"

https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala#L128

### CODE REVIEW

**Comment by wsargent** [comment]
Should also backport to 2.4.x

File: documentation/manual/working/javaGuide/main/ws/JavaOAuth.md [hunk 1]
```diff
@@ -50,4 +50,4 @@ controller:
 
 @[ws-oauth-controller](code/javaguide/ws/controllers/Twitter.java)
 
-> **NOTE**: OAuth does not provide any protection against [MITM attacks](http://en.wikipedia.org/wiki/Man-in-the-middle_attack).  This example shows the OAuth token and secret stored in a session cookie -- for the best security, always use HTTPS with `play.http.session.cookie.secure=true` defined.
+> **NOTE**: OAuth does not provide any protection against [MITM attacks](http://en.wikipedia.org/wiki/Man-in-the-middle_attack).  This example shows the OAuth token and secret stored in a session cookie -- for the best security, always use HTTPS with `play.http.session.secure=true` defined.
```
Comment (marcospereira) [inline documentation/manual/working/javaGuide/main/ws/JavaOAuth.md]: I think this note should be consistent with `documentation/manual/working/scalaGuide/main/ws/ScalaOAuth.md` note.

**Comment by jroper** [comment]
Backported to 2.4.x: 02fa22b

**Comment by philipwhiuk** [comment]
1. Why isn't this true by default - applications should present a secure by default approach - minimal configuration for a secure system. The lesson we were supposed to learn from papers like: http://www.cs.utexas.edu/~shmat/shmat_ccs12.pdf is that leaving odd bits of code to be reconfigured by the user in order to get a secure set-up leads to insecurity.
2. Has this now changed name to `play.http.cookies.strict` ?

**Comment by wsargent** [comment]
@philipwhiuk please see https://github.com/playframework/playframework/blob/master/templates/play-common/conf/application.conf#L127

**Comment by wsargent** [comment]
@philipwhiuk You may also like https://tersesystems.com/2014/01/13/fixing-the-most-dangerous-code-in-the-world/

**Comment by philipwhiuk** [comment]
@wsargent Thanks. Yes. I know the documentation refers to session.secure. The actual code, in master on the mentioned line as noted in comment 1, is looking for: `play.http.cookies.strict`. I would say that one of them is wrong.

**Comment by wsargent** [comment]
Oh -- the link is wrong.  `play.http.cookies.strict` is true by default and shouldn't be changed: https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/resources/reference.conf#L120

**Comment by jroper** [comment]
The strict flag was added after a security vulnerability was found that allowed injecting cookies in certain (very specific) situations.  The problem with the fix was that it may have broken sites that used invalid cookies, and this isn't something that you always have complete control over since things like Google analytics add their own cookies, so we put an option there to turn the strict parsing off.  It's documented in the [reference.conf](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/resources/reference.conf#L120), it wasn't documented anywhere else because it really should be rare that anyone would ever turn it off, it's just an escape hatch if ever needed.

As for `secure` being true by default, since by default Play runs HTTP by default, not HTTPS, using secure cookies by default would mean Play sessions would not work with Play's default configuration, the first thing you'd have to do when starting working with would be turn off `secure`.  A non working out of the box configuration is not really an option.

**Comment by wsargent** [comment]
> As for secure being true by default, since by default Play runs HTTP by default, not HTTPS, using secure cookies by default would mean Play sessions would not work with Play's default configuration, the first thing you'd have to do when starting working with would be turn off secure. A non working out of the box configuration is not really an option.

The work around for 2.5.x is to have all the secure / httpOnly options explicitly referenced and commented out, so the only thing you need to do is uncomment them.  

A better option may be to have global substitution `${secureFlag}` option in HOCON, and then turn on all the options at once.

**Pull request merged**